File: 76-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if(s == null || t == null || s.length() < t.length()) { return ""; }    
        
        Map<Character, Integer> freqMap = new HashMap<>();
        for(char c : t.toCharArray()) {
            freqMap.put(c, freqMap.getOrDefault(c, 0) + 1);
        }
        
        int start = 0;
        int minLen = Integer.MAX_VALUE;
        int matches = 0;
        
        int left = 0;
        for(int right = 0; right < s.length(); right++) {
            char rightChar = s.charAt(right);
            if(freqMap.containsKey(rightChar)) {
                freqMap.put(rightChar, freqMap.get(rightChar) - 1);
                if(freqMap.get(rightChar) >= 0) {
                    matches++;
                }                
            }
            
            while(matches == t.length()) {
                int currentLen = right - left + 1;
                if(currentLen < minLen) {
                    minLen = currentLen;
                    start = left;
                }
                
                char leftChar = s.charAt(left);
                if(freqMap.containsKey(leftChar)) {
                    if(freqMap.get(leftChar) == 0) {
                        matches--;
                    }
                    
                    freqMap.put(leftChar, freqMap.get(leftChar) + 1);
                }
                
                left++;
            }
        }
        
        return minLen == Integer.MAX_VALUE? "" : s.substring(start, start + minLen);
    }
}

File: 76_Minimum_Window_Substring.py
# Sliding window
I use a sliding window with two counters: required_count tells me how many of each character I need, and window_count tracks what the current window contains.
required_types is the number of distinct characters needed, and satisfied_types is how many of those are currently satisfied.

I expand the window with the right pointer until all required types are satisfied.
Once the window is valid, I shrink from the left to find the smallest possible valid window, updating the best answer.
This guarantees O(n) time because each pointer only moves forward.
  
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        required_count = Counter(t)          # required frequencies of t
        window_count = Counter()             # current window frequencies

        required_types = len(required_count) # number of distinct chars required
        satisfied_types = 0                  # number of distinct chars satisfied

        left = 0
        best_len = float('inf')
        best_start = -1

        for right in range(len(s)):
            ch = s[right]

            # expand window
            if ch in required_count:
                window_count[ch] += 1
                if window_count[ch] == required_count[ch]:
                    satisfied_types += 1

            # shrink window
            while satisfied_types == required_types:
                curr_len = right - left + 1
                if curr_len < best_len:
                    best_len = curr_len
                    best_start = left

                left_char = s[left]
                if left_char in required_count:
                    if window_count[left_char] == required_count[left_char]:
                        satisfied_types -= 1
                    window_count[left_char] -= 1

                left += 1

        return "" if best_start == -1 else s[best_start : best_start + best_len]

